doc/dev/static_type_checking_cheat_sheet.md

```diff
@@ -0,0 +1,410 @@
+# Static Type Checking Cheat Sheet
+
+This cheat sheet details guidance for typing as it relates to the Python SDK. Use your own judgment to achieve the best balance of clarity and flexibility for your Python client library. 
+
+### General guidance
+
+- Do provide type annotations (per PEP 484) to public APIs in the client library.
+- You do not need to type annotate all functions in a client library, but you should provide type hints where unit tests are worth writing or where type annotations will assist in understanding of the code.
```

Comment:
Maybe further clarify that this is referring to internal implementation functions.